src/ucp/core/ucp_ep.c

```diff
@@ -2330,3 +2333,75 @@ void ucp_ep_do_keepalive(ucp_ep_h ep, ucp_lane_map_t *lane_map)
         }
     }
 }
+
+static void ucp_ep_req_purge(ucp_ep_h ucp_ep, ucp_request_t *req,
+                             ucs_status_t status, int recursive)
+{
+    if (ucp_worker_get_request_by_id(ucp_ep->worker, req->req_id.local) != NULL) {
+        ucp_worker_del_request_id(ucp_ep->worker, req);
+    }
+
+    if (req->flags & (UCP_REQUEST_FLAG_SEND_AM | UCP_REQUEST_FLAG_SEND_TAG)) {
+        ucs_assert(req->super_req == NULL);
+        ucs_assert(!(req->flags & UCP_REQUEST_FLAG_OFFLOADED));
+        ucp_request_complete_and_dereg_send(req, status);
+    } else if (req->flags & (UCP_REQUEST_FLAG_RECV_AM | UCP_REQUEST_FLAG_RECV_TAG)) {
+        ucs_assert(req->super_req == NULL);
+        ucp_request_complete_and_dereg_recv_rndv(req, status);
+    } else if (req->flags & UCP_REQUEST_FLAG_RNDV_FRAG) {
+        ucs_assert(req->super_req != NULL);
+        ucs_assert(recursive); /* mustn't be directly contained in an EP list
+                                * of tracking requests */
+
+        /* it means that purging started from a request responsible for sending RTR,
+         * so a request is responsible for copying data from staging buffer and it
+         * uses a receive part of a request */
+        req->super_req->recv.remaining -= req->recv.length;
+        if (req->super_req->recv.remaining == 0) {
+            ucp_ep_req_purge(ucp_ep, req->super_req, status, 1);
+        }
```

Comment:
seems like we can lost actual status here if last chunk of data will be completed successfully, so successfully completed request will contain invalid data in the buffer
need to store first NOT_OK status and use it, see logic in `uct_completion_update_status`

Reply:
they will be completed with the status provided by error callback

Reply:
why? we have to wait all completions from transports, so what if request has 1 posted operation (or not handled successful completion) at this point?

Reply:
> why? we have to wait all completions from transports

correct, we do it

requests that are kept in this list are not scheduled on a UCT transport (am/rma/amo/pending)
e.g. if we have fragmented GET, we have super a GET request and multiple (let's assume 5 requests for simplicity) that do GET operation. let's also assume that 3 requests are already completed a GET operation and waiting for ATS, and 2 requests are scheduled on UCT transport and waiting for GET operation completion. At this point, the switch port is down, so, the sender is inaccessible. Uncompleted UCT operations report completion with error for 2 requests. And then we error handling callback is called and we do purging of UCP requests that are under tracking. We will update `recv.remaining` of super receive requests and once the last UCP request is purged, `recv.remaining` is 0, and we complete super request with error.

Reply:
what if err callback was triggered by another request on other lane?

Reply:
> what if err callback was triggered by another request on other lane?

2 requests from example will be completed with success and added to the list of tracking requests (to wait for ATS).
anyway, UCT EP -> UCP EP lookup will be done during error handling and requests from tracking list on UCP EP will be purged

Reply:
successful completions can be polled out later and complete request with status UCS_OK

Reply:
@evgeny-leksikov I provided the wrong example in the comment above.
let's assume that we are a receiver - and we requesting fragmented PUT, we send several RTRs (i.e. RTR per fragment). each RTR request is added to EP tracking list. we complete RTR request upon receiving ATP packet.
I agree we should update the status for the parent request. Will do it. It is needed only for fragmented operations. Other RNDV operations do it (but need to do it from UCT completion callback as well - #6151)